packages/frontend/app/clipboard-history/page.tsx

```diff
@@ -3,20 +3,161 @@
 
 export default function ClipboardHistoryPage() {
+    const {close} = useWindowControls();
+    const [query, setQuery] = useState<string>('');
+    const [typeFilter, setTypeFilter] = useState<'' | ClipboardType>('');
+    const [selectedIndex, setSelectedIndex] = useState<number>(0);
+    const {isLoading, settings} = useSettings<ClipboardSettings>('clipboard');
+
+    const {items, history} = useMemo<{items: ClipboardItems, history: ClipboardHistory}>(() => ({
+        items: isLoading ? {} : settings.clipboardItems,
+        history: isLoading ? [] as ClipboardHistory : settings.clipboardHistory,
+    }), [isLoading, settings]);
+
+    const selectedItem = useMemo<ClipboardItem>(() => items[history[selectedIndex]] ?? undefined, [selectedIndex, items, history]);
+
+    const updateQuery = useCallback((query: string) => {
+        setQuery(query);
+        setSelectedIndex(0);
+    }, [setQuery]);
+
+    const updateTypeFilter = useCallback((type: '' | ClipboardType) => {
+        setTypeFilter(type);
+        setSelectedIndex(0);
+    }, [setQuery]);
+
+    const searchQuery = useMemo(() => query.toLowerCase(), [query]);
+
+    const searchableItems = useMemo<SearchableItem[]>(() => {
+        return history.map((hash: ClipboardItemHash) => {
+            const item = items[hash] ?? undefined;
+            if (item) {
+                const {type, value} = item;
+                switch (type) {
+                    case "text":
+                    case "colour":
+                        return {
+                            hash,
+                            type,
+                            text: value.toLowerCase(),
+                        }
+                    case "html":
+                    case "url":
+                    case "path":
+                        const {metadata: { text }} = item
+                        return {
+                            hash,
+                            type,
+                            text: text.toLowerCase(),
+                        }
+                    case "image":
+                        return {
+                            hash,
+                            type,
+                            text: `image`
+                        }
+                }
+            }
+        })
+    }, [items, history]);
+
+    const filteredHistory = useMemo<ClipboardHash[]>(() => {
+        const filtered = [];
+
+        searchableItems.forEach(({ text, type, hash }) => {
+            const textMatch = text.includes(searchQuery);
+
+            const typeSearch = typeFilter !== '';
+            const typeMatch = typeSearch ? typeFilter === type : true;
+
+            if (textMatch && typeMatch) {
+                filtered.push(hash);
+            }
+        })
+
+        return filtered;
+    }, [searchableItems, searchQuery, typeFilter]);
+
+    const selectNext = useCallback(() => {
+        const newSelectedIndex = selectedIndex + 1;
+        const hasNew = !!filteredHistory[newSelectedIndex];
```

Comment:
@coderabbitai can you explain why filteredHistory is always an empty string (it's default state), even though there should be items in? Guess I messed up something with react hooks / dependencies.

Reply:
Could you please re-check that, I'm not sure if that breaks the react hook system?